p11-kit/list-mechanisms.c

```diff
@@ -0,0 +1,303 @@
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ * Author: Alexander Sosedkin <asosedkin@redhat.com>
+ */
+
+#include "config.h"
+
+#include "constants.h"
+#include "debug.h"
+#include "iter.h"
+#include "message.h"
+#include "pkcs11.h"
+#include "print.h"
+#include "tool.h"
+#include "uri.h"
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#ifdef ENABLE_NLS
+#include <libintl.h>
+#define _(x) dgettext(PACKAGE_NAME, x)
+#else
+#define _(x) (x)
+#endif
+
+int
+p11_kit_list_mechanisms (int argc,
+		         char *argv[]);
+
+static void
+print_mechanism_with_info(CK_MECHANISM_TYPE mechanism,
+		          CK_MECHANISM_INFO info)
+{
+	const char *mechanism_nick = NULL;
+
+	mechanism_nick = p11_constant_nick (p11_constant_mechanisms, mechanism);
+	if (mechanism_nick == NULL)
+		printf ("0x%lX (unknown):", mechanism);
+	else
+		printf ("%s:", mechanism_nick);
+
+	if (info.flags & CKF_HW)
+		printf (" hw");
+	if (info.flags & CKF_MESSAGE_ENCRYPT)
+		printf (" message-encrypt");
+	if (info.flags & CKF_MESSAGE_DECRYPT)
+		printf (" message-decrypt");
+	if (info.flags & CKF_MESSAGE_SIGN)
+		printf (" message-sign");
+	if (info.flags & CKF_MESSAGE_VERIFY)
+		printf (" message-verify");
+	if (info.flags & CKF_MULTI_MESSAGE)
+		printf (" multi-message");
+	if (info.flags & CKF_FIND_OBJECTS)
+		printf (" find-objects");
+	if (info.flags & CKF_ENCRYPT)
+		printf (" encrypt");
+	if (info.flags & CKF_DECRYPT)
+		printf (" decrypt");
+	if (info.flags & CKF_DIGEST)
+		printf (" digest");
+	if (info.flags & CKF_SIGN)
+		printf (" sign");
+	if (info.flags & CKF_SIGN_RECOVER)
+		printf (" sign-recover");
+	if (info.flags & CKF_VERIFY)
+		printf (" verify");
+	if (info.flags & CKF_SIGN_RECOVER)
+		printf (" verify-recover");
+	if (info.flags & CKF_GENERATE)
+		printf (" generate");
+	if (info.flags & CKF_GENERATE_KEY_PAIR)
+		printf (" generate-key-pair");
+	if (info.flags & CKF_WRAP)
+		printf (" wrap");
+	if (info.flags & CKF_UNWRAP)
+		printf (" unwrap");
+	if (info.flags & CKF_DERIVE)
+		printf (" derive");
+	if (info.flags & CKF_EXTENSION)
+		printf (" extension");
+
+	if (info.ulMaxKeySize)
+		printf (" key-size=%ld-%ld", info.ulMinKeySize, info.ulMaxKeySize);
+	printf ("\n");
+}
+
+static int
+list_mechanisms (const char *token_str)
+{
+	int ret = 1;
+	CK_FUNCTION_LIST **modules = NULL;
+	CK_FUNCTION_LIST *module = NULL;
+	P11KitUri *uri = NULL;
+	P11KitIter *iter = NULL;
+	CK_SESSION_HANDLE session = 0;
+	CK_SLOT_ID slot = 0;
+	CK_MECHANISM_INFO mechanism_info;
+	CK_MECHANISM_TYPE_PTR mechanisms = NULL;
+	CK_MECHANISM_TYPE mechanism = CKP_INVALID_ID;
```

Comment:
Type mismatch: `CKP_INVALID_ID` is defined as CK_PROFILE_ID

Reply:
side-stepped